capellambse/extensions/validation/_validate.py

```diff
@@ -0,0 +1,263 @@
+    value: bool
+
+
+Validator = cabc.Callable[
+    [c.GenericElement], bool
+]  # | Callable [[Metric], Result]
+Type = t.Union[
+    str, type[c.GenericElement]
+]  # | type[Metric] | type[AggregateMetric]
+"""Type to be used in capellambse.model.MelodyModel.search(type) or metric
+function the value of which will be validated by the rule."""
+
+
+@dataclasses.dataclass(frozen=True, eq=True)
+class Rule:
+    """Common (generic) class representing any validation rule."""
+
+    id: str
+    name: str
+    rationale: str
+    actions: list[str]
+    validator: Validator
+    hyperlink_further_reading: str | None = None
+    """The actual function will be assigned to this attribute, so that we do
+    not need to have 100 classes for 100 rules."""
+
+    def __call__(self, obj: c.GenericElement) -> bool:  # | Metric
+        return self.validator(obj)
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+
+class Results(dict[Rule, dict[helpers.UUIDString, Result]]):
+    """A validation rule result mapping."""
+
+    def __getitem__(self, key: Rule | str) -> dict[helpers.UUIDString, Result]:
+        if isinstance(key, str):
+            for rule in self:
+                if rule.id == key:
+                    key = rule
+                    break
+            else:
+                raise KeyError(f"No rule with id: {key!r} found")
+
+        return super().__getitem__(key)
+
+    def by_uuid(self, uuid: helpers.UUIDString) -> Results:
+        """Filter the validation results by ``uuid``."""
+        return Results(
+            {
+                rule: {uid: r for uid, r in res.items() if uid == uuid}
+                for rule, res in self.items()
+            }
+        )
+
+    def by_category(self, category: Category) -> Results:
+        """Filter the validation results by ``category``."""
+        return Results(
+            {
+                rule: {
+                    uid: r for uid, r in res.items() if r.category == category
+                }
+                for rule, res in self.items()
+            }
+        )
+
+    def by_value(self, value: bool) -> Results:
+        """Filter the validation results by ``value``."""
+        return Results(
+            {
+                rule: {uid: r for uid, r in res.items() if r.value == value}
+                for rule, res in self.items()
+            }
+        )
+
+    def setdefault(
+        self,
+        key: t.Any,
+        default: dict[helpers.UUIDString, Result] | None = None,
+    ) -> dict[helpers.UUIDString, Result]:
+        if not isinstance(default, dict):
+            raise ValueError("Only a result dictionary is accepted")
+        return super().setdefault(key, default)
+
+
+VALIDATION_RULES: dict[Category, dict[Type, list[Rule]]] = {
+    Category.REQUIRED: {},
+    Category.RECOMMENDED: {},
+    Category.SUGGESTED: {},
+}
+VALIDATION_RESULTS = Results()
+
+
+def register_rule(
+    category: Category,
+    type: Type,
```

Comment:
I'm convinced, @MS34449, that we can generalize rules and actions to keep it DRY yet helpful. The engine shall support assignment of one rule to multiple object types to enable that kind of generalization. If somebody wants to repeat rules like "an object of significance shall have a summary or description so that model users could understand what that object represents" for every object of significance like a Component, Function, Capability, ... this is totaly fine and they can but we should not limit those users who can create DRY rules. Hence the code shall be improved as per @Wuestengecko's comment.